src/components/Admin/Trainer/Trainer.js

```diff
@@ -0,0 +1,267 @@
+			this.state.item,
+			this.state.burstCount
+		);
+	};
+
+	addFromFile = () => {
+		if (this.files.current.files.length < 1) {
+			this.setStatus('Please choose some files first.');
+			return;
+		}
+
+		this.setState({busy: true});
+
+		this.images = [];
+		this.fileIndex = 0;
+		let index = 0;
+		const reader = new FileReader();
+
+		reader.addEventListener('load', result => {
+			this.images.push(result.target.result);
+			if (index < this.files.current.files.length) {
+				reader.readAsDataURL(this.files.current.files[index++]);
+			} else {
+				this.model.addExample(
+					() => this.images[this.fileIndex++],
+					this.captureFromFile,
+					this.state.item,
+					this.files.current.files.length
+				);
+			}
+		});
+
+		reader.readAsDataURL(this.files.current.files[index++]);
+	};
+
+	train = () => {
+		this.setState({busy: true});
+
+		const {
+			hiddenUnits,
+			batchSizeFraction,
+			learningRate,
+			epochs,
+			setSize,
+			randomness,
+			since
+		} = this.state;
+
+		this.model.train(
+			hiddenUnits,
+			batchSizeFraction,
+			learningRate,
+			epochs,
+			setSize,
+			randomness,
+			Date.parse(since)
+		);
+	};
+
+	predict = () => {
+		if (this.webcam.current.webcam.current) {
+			this.setState({busy: true});
+			this.model.predict(
+				this.capture(this.webcam.current.webcam.current.video)
+			);
+		} else {
+			this.setStatus('Please connect a camera.');
+		}
+	};
+
+	getName = item => {
+		return item.name + (item.qualifier ? ` (${item.qualifier})` : '');
+	};
+
+	setStatus = status => {
+		this.setState({status});
+		if (status.indexOf('...') === -1) this.setState({busy: false});
+	};
+
+	render() {
+		const {
+			learningRate,
+			batchSizeFraction,
+			epochs,
+			hiddenUnits,
+			setSize,
+			randomness,
+			since,
+			busy,
+			status,
+			item,
+			burstCount
+		} = this.state;
+
+		return (
+			<div>
+				<div className="col" style={{textAlign: 'center'}}>
+					<span id="status-text">{status}</span>
+					<br />
+					<WebcamCapture
+						setPrediction={() => {}}
+						history={{push: () => {}}}
+						ref={this.webcam}
+					/>
+					<br />
+					<ItemSelector
+						item={item}
+						items={Object.values(this.model.items)}
+						setItem={item => this.setState({item})}
+						disabled={busy}
+					/>
+					<br />
+					<label>Burst Count:</label>
+					<input
+						type="text"
+						value={burstCount}
+						size={3}
+						maxLength={3}
+						disabled={busy}
+						onChange={e =>
+							this.setState({burstCount: e.target.value})
+						}
+					/>
+					<button
+						onClick={() =>
+							this.addExample(
+								this.webcam.current.webcam.current.video
+							)
+						}
+						disabled={busy || !this.webcam.current.webcam.current}>
+						Add From Camera
+					</button>
+					<br />
+					<input
+						type="file"
+						multiple
+						ref={this.files}
+						disabled={busy}
+					/>
+					<button onClick={this.addFromFile} disabled={busy}>
+						Add From File
```

Comment:
If the Admin accidentally clicks add from file with no file attached, all buttons freeze. It would be annoying if lots of images have been loaded this happens